contracts/staking/StakeHolder.sol

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2024
+// SPDX-License-Identifier: Apache 2
+pragma solidity ^0.8.19;
+
+import {UUPSUpgradeable} from "openzeppelin-contracts-upgradeable-4.9.3/proxy/utils/UUPSUpgradeable.sol";
+import {AccessControlEnumerableUpgradeable} from "openzeppelin-contracts-upgradeable-4.9.3/access/AccessControlEnumerableUpgradeable.sol";
+
+/// @notice Struct to combine an account and an amount.
+struct AccountAmount {
+    address account;
+    uint256 amount;
+}
+
+/**
+ * @title StakeHolder: allows anyone to stake any amount of native IMX and to then remove all or part of that stake.
+ * @dev The StakeHolder contract is designed to be upgradeable.
+ */
+contract StakeHolder is AccessControlEnumerableUpgradeable, UUPSUpgradeable {
+    /// @notice Error: Attempting to upgrade contract storage to version 0.
+    error CanNotUpgradeToLowerOrSameVersion(uint256 _storageVersion);
+
+    /// @notice Error: Attempting to renounce the last role admin / default admin.
+    error MustHaveOneRoleAdmin();
+
+    /// @notice Error: Attempting to stake with zero value.
+    error MustStakeMoreThanZero();
+
+    /// @notice Error: Attempting to distribute zero value.
+    error MustDistributeMoreThanZero();
+
+    /// @notice Error: Attempting to unstake amount greater than the balance.
+    error UnstakeAmountExceedsBalance(uint256 _amountToUnstake, uint256 _currentStake);
+
+    /// @notice Error: The sum of all amounts to distribute did not equal msg.value of the distribute transaction.
+    error DistributionAmountsDoNotMatchTotal(uint256 _msgValue, uint256 _calculatedTotalDistribution);
+
+    /// @notice Error: Distributions can only be made to accounts that have staked.
+    error AttemptToDistributeToNewAccount(address _account, uint256 _amount);
+
+    /// @notice Event when an amount has been staked or when an amount is distributed to an account.
+    event StakeAdded(address _staker, uint256 _amountAdded, uint256 _newBalance);
+
+    /// @notice Event when an amount has been unstaked.
+    event StakeRemoved(address _staker, uint256 _amountRemoved, uint256 _newBalance);
+
+    /// @notice Event summarising a distribution. There will also be one StakeAdded event for each recipient.
+    event Distributed(address _distributor, uint256 _totalDistribution, uint256 _numRecipients);
+
+    /// @notice Only UPGRADE_ROLE can upgrade the contract
+    bytes32 public constant UPGRADE_ROLE = bytes32("UPGRADE_ROLE");
+
+    /// @notice Version 0 version number
+    uint256 private constant _VERSION0 = 0;
+
+    /// @notice Holds staking information for a single staker.
+    struct StakeInfo {
+        /// @notice Amount of stake.
+        uint256 stake;
+        /// @notice True if this account has ever staked.
+        bool hasStaked;
```

Comment:
just reading the variable name `hasStaked` seemed to me like 'if the user has currently staked or not' only after reading the comment it made it clear for me. What is the purpose of knowing `hasStaked`? Is it more like history? When the user unstakes completely does this struct become `{stake: 0, hasStaked: true}`?

Reply:
>When the user unstakes completely does this struct become {stake: 0, hasStaked: true}?
Correct

Reply:
The purpose of hasStaked is to ensure the stakers array has a list of unique addresses: that is, there are no duplicate addresses. If a staker completely unstakes (stake == 0) and then restakes (stake != 0), having hasStaked allows the _addStake function on line 230 to know that the staker's address is already in the stakers array.